src/cpu/core_dyn_x86.cpp

```diff
@@ -248,13 +257,34 @@ static void dyn_restoreregister(DynReg * src_reg, DynReg * dst_reg) {
 
 #include "core_dyn_x86/decoder.h"
 
-static void copy_dh_fpu_to_normal()
+#	if defined(X86_DYNFPU_DH_ENABLED)
+
+static void maybe_copy_host_fpu_to_dh()
 {
-	FPU_SetTag(dyn_dh_fpu.state.tag);
-	FPU_SetCW(dyn_dh_fpu.state.cw);
-	FPU_SetSW(dyn_dh_fpu.state.sw);
+	if (dyn_dh_fpu.state_used) {
+		gen_dh_fpu_save();
+		dyn_dh_fpu.state_copied_to_normal = false;
+	}
+}
+
+static void maybe_copy_dh_fpu_to_normal()
+{
+	if (dyn_dh_fpu.state_copied_to_normal) {
```

Comment:
I have a PR in the X repo that utilises similar idea.  But I perform the copies at the 2 choke points:
* entering normal core
* entering dynamic core

Reply:
Thanks for the heads up, @finalpatch 👍